src/CountryCode.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\I18n\PhoneNumber;
+
+use Laminas\I18n\PhoneNumber\Exception\InvalidArgumentException;
+use Locale;
+
+use function preg_match;
+use function sprintf;
+use function strtoupper;
+
+/**
+ * @psalm-immutable
+ */
+final class CountryCode
+{
+    /**
+     * @readonly
+     * @var non-empty-string
+     */
+    private string $code;
+
+    /** @param non-empty-string $code */
+    private function __construct(string $code)
+    {
+        $this->code = $code;
+    }
+
+    /** @return non-empty-string */
+    public function toString(): string
+    {
+        return $this->code;
+    }
+
+    public function equals(self $other): bool
+    {
+        return $this->code === $other->code;
+    }
+
+    /** @param non-empty-string $code */
+    public static function fromString(string $code): self
+    {
+        $code = strtoupper($code);
+        if (! preg_match('/^[A-Z]{2}$/', $code)) {
+            throw new InvalidArgumentException('Country codes should be 2 letter ISO 3166 strings');
+        }
+
+        $displayName = Locale::getDisplayRegion('-' . $code, 'GB');
+        if ($displayName === '' || $displayName === 'Unknown Region') {
+            throw new InvalidArgumentException(sprintf(
+                'The country code "%s" does not correspond to a known country',
+                $code
+            ));
+        }
+
+        return new self($code);
+    }
+
+    /** @param non-empty-string $locale */
+    public static function fromLocale(string $locale): self
```

Comment:
I am a little confused with the method names because `fromString()` creates a the value object from a string. The name of the method `fromLocale()` also creates the object from a string, but not from the PHP's Locale `Locale` as the name of the method suggests.
The method signatures are clear, but the name and their handling confuse me a little.

Reply:
IMO, `fromString()` is fine, but I think @froschdesign means `fromLocale()` should become `fromLocaleString()` for clarity.